File: ScaleCalculator/ScaleCalculator/SingleVanishingPoint_simplified.cpp
#include "SingleVanishingPoint_simplified.h"

SingleVanishingPoint_simplified::SingleVanishingPoint_simplified(std::vector<Point> corners, LineSegment measure, Point vPoint):Region(corners)
{
	this->measure = measure;
	this->vPoint = vPoint;
}

Point SingleVanishingPoint_simplified::buildEndpoint_approximate(Point start, LineSegment direction)
{
	LineSegment directionProjection(
		CalculateIntersection(BuildPerpendicular(Line(measure), direction.start), Line(measure)),
		CalculateIntersection(BuildPerpendicular(Line(measure), direction.end), Line(measure)), 0);
	double projectionRatio = directionProjection.GetPixelLength() / direction.GetPixelLength();
	directionProjection.length = direction.length * projectionRatio;

	double pixelLength = (directionProjection.length * (measure.GetPixelLength() / measure.length))/ projectionRatio;

	LineSegment measuredSegment = direction.BuildCroppedSegmentFromPoint(CalculateIntersection(Line(measure),Line(vPoint, start)), pixelLength);

	Line dirLine(direction.BuildCroppedSegmentFromPoint(start, pixelLength));

	return CalculateIntersection(dirLine, Line(measuredSegment.end, vPoint));
}

void SingleVanishingPoint_simplified::Deserialize(const json& segment)
{
	this->measure.Deserialize(segment["measure"]);
	this->vPoint.Deserialize(segment["vPoint"]);
}

File: MetadataGenerator/CPPclasses/SingleVanishingPoint_simplified.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CPPclasses
{
    public class SingleVanishingPoint_simplified : Region
    {
        public LineSegment measure { get; set; }
        public Point vPoint { get; set; }
        /// <summary>
        /// Create Region with Single Vanishing Point.
        /// </summary>
        /// <param name="corners">Corners, that limit Region</param>
        /// <param name="measure">!!!Only horizontal and vertical lines!!!  line on a plane, with known size.</param>
        /// <param name="vPoint">Vanishing point</param>
        public SingleVanishingPoint_simplified(List<Point> corners, LineSegment measure, Point vPoint) : base(corners)
        {
            this.measure = measure;
            this.vPoint = vPoint;
        }
        public SingleVanishingPoint_simplified() { }

        // Унаследовано через Region
        public override Point buildEndpoint_approximate(Point start, LineSegment direction)
        {
            LineSegment directionProjection = new(
                                                    Measures.CalculateIntersection(Measures.BuildPerpendicular(new Line(measure), direction.start), new Line(measure)),
                                                    Measures.CalculateIntersection(Measures.BuildPerpendicular(new Line(measure), direction.end), new Line(measure)), 0);
            double projectionRatio = directionProjection.GetPixelLength() / direction.GetPixelLength();
            directionProjection.length = direction.length * projectionRatio;

            double pixelLength = (directionProjection.length * (measure.GetPixelLength() / measure.length)) / projectionRatio;

            LineSegment measuredSegment = direction.BuildCroppedSegmentFromPoint(Measures.CalculateIntersection(new Line(measure), new Line(vPoint, start)), pixelLength);

            Line dirLine = new(direction.BuildCroppedSegmentFromPoint(start, pixelLength));

            return Measures.CalculateIntersection(dirLine, new Line(measuredSegment.end, vPoint));
        }
    }
}
